File: apps/macos/Sources/Persistence/AppConfiguration.swift
import CoreGraphics
import Foundation

/// Energy rules the user can opt out of. Nothing here changes behaviour
/// silently: every rule maps to one switch in Settings.
nonisolated struct EnergyPreferences: Codable, Hashable, Sendable {
    var pauseInLowPowerMode: Bool = true
    var pauseOnHighThermalState: Bool = true
    var pauseWhenScreenLocked: Bool = true

    init(
        pauseInLowPowerMode: Bool = true,
        pauseOnHighThermalState: Bool = true,
        pauseWhenScreenLocked: Bool = true
    ) {
        self.pauseInLowPowerMode = pauseInLowPowerMode
        self.pauseOnHighThermalState = pauseOnHighThermalState
        self.pauseWhenScreenLocked = pauseWhenScreenLocked
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        pauseInLowPowerMode = try container.decodeIfPresent(Bool.self, forKey: .pauseInLowPowerMode) ?? true
        pauseOnHighThermalState = try container.decodeIfPresent(Bool.self, forKey: .pauseOnHighThermalState) ?? true
        pauseWhenScreenLocked = try container.decodeIfPresent(Bool.self, forKey: .pauseWhenScreenLocked) ?? true
    }
}

/// Everything the app restores after a relaunch. Transient objects
/// (`AVPlayer`, `NSScreen`, window references) are deliberately absent.
nonisolated struct AppConfiguration: Codable, Hashable, Sendable {
    static let currentVersion = 1

    var version: Int = AppConfiguration.currentVersion
    var library: [WallpaperRecord] = []
    var displayConfigurations: [DisplayWallpaperConfiguration] = []
    /// Whether wallpapers were running when the app last quit.
    var isWallpaperEnabled: Bool = false
    var energyPreferences: EnergyPreferences = EnergyPreferences()
    /// The desktop picture each display had before the app replaced it, keyed by
    /// display ID. Kept so switching the option off restores the user's own
    /// wallpaper instead of stranding them with a video frame.
    var previousDesktopPictures: [String: String] = [:]

    init() {}

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        version = try container.decodeIfPresent(Int.self, forKey: .version) ?? Self.currentVersion
        library = try container.decodeIfPresent([WallpaperRecord].self, forKey: .library) ?? []
        displayConfigurations = try container.decodeIfPresent(
            [DisplayWallpaperConfiguration].self,
            forKey: .displayConfigurations
        ) ?? []
        isWallpaperEnabled = try container.decodeIfPresent(Bool.self, forKey: .isWallpaperEnabled) ?? false
        energyPreferences = try container.decodeIfPresent(
            EnergyPreferences.self,
            forKey: .energyPreferences
        ) ?? EnergyPreferences()
        previousDesktopPictures = try container.decodeIfPresent(
            [String: String].self,
            forKey: .previousDesktopPictures
        ) ?? [:]
    }

    func wallpaper(withID id: UUID?) -> WallpaperRecord? {
        guard let id else { return nil }
        return library.first { $0.id == id }
    }

    func configuration(forDisplay displayID: CGDirectDisplayID) -> DisplayWallpaperConfiguration? {
        displayConfigurations.first { $0.displayID == displayID }
    }

    /// Inserts or replaces one display assignment, keeping the collection
    /// free of duplicates for the same display.
    mutating func setConfiguration(_ configuration: DisplayWallpaperConfiguration) {
        if let index = displayConfigurations.firstIndex(where: { $0.displayID == configuration.displayID }) {
            displayConfigurations[index] = configuration
        } else {
            displayConfigurations.append(configuration)
        }
    }

    /// Drops assignments that point at a wallpaper that is no longer in the
    /// library, so a deleted video cannot resurrect a broken session.
    mutating func removeAssignments(ofWallpaper id: UUID) {
        for index in displayConfigurations.indices where displayConfigurations[index].wallpaperID == id {
            displayConfigurations[index].wallpaperID = nil
        }
    }
}

File: apps/windows/src/Wallpache.App/Persistence/AppConfiguration.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.Json.Serialization;
using Wallpache.App.Displays;
using Wallpache.App.Library;

namespace Wallpache.App.Persistence;

/// <summary>
/// Energy rules the user can opt out of. Nothing here changes behaviour
/// silently: every rule maps to one switch in Settings.
/// </summary>
public sealed class EnergyPreferences
{
    [JsonPropertyName("pauseInBatterySaver")]
    public bool PauseInBatterySaver { get; set; } = true;

    [JsonPropertyName("pauseOnBatteryPower")]
    public bool PauseOnBatteryPower { get; set; } = true;

    [JsonPropertyName("pauseWhenScreenLocked")]
    public bool PauseWhenScreenLocked { get; set; } = true;

    public EnergyPreferences Copy() => new()
    {
        PauseInBatterySaver = PauseInBatterySaver,
        PauseOnBatteryPower = PauseOnBatteryPower,
        PauseWhenScreenLocked = PauseWhenScreenLocked
    };

    public bool Matches(EnergyPreferences other) =>
        PauseInBatterySaver == other.PauseInBatterySaver
        && PauseOnBatteryPower == other.PauseOnBatteryPower
        && PauseWhenScreenLocked == other.PauseWhenScreenLocked;
}

/// <summary>
/// Everything the app restores after a relaunch. Transient objects (players,
/// monitor handles, window handles) are deliberately absent.
///
/// Every property tolerates a missing value so that a settings file written by
/// an older build never wipes a user's assignments.
/// </summary>
public sealed class AppConfiguration
{
    public const int CurrentVersion = 1;

    [JsonPropertyName("version")]
    public int Version { get; set; } = CurrentVersion;

    [JsonPropertyName("wallpapers")]
    public List<WallpaperRecord> Library { get; set; } = [];

    [JsonPropertyName("displays")]
    public List<DisplayWallpaperConfiguration> DisplayConfigurations { get; set; } = [];

    /// <summary>Whether wallpapers were running when the app last quit.</summary>
    [JsonPropertyName("isWallpaperEnabled")]
    public bool IsWallpaperEnabled { get; set; }

    [JsonPropertyName("launchAtSignIn")]
    public bool LaunchAtSignIn { get; set; }

    [JsonPropertyName("energyPreferences")]
    public EnergyPreferences EnergyPreferences { get; set; } = new();

    /// <summary>
    /// The desktop picture each display had before the app replaced it, keyed by
    /// display id. Kept so switching the option off restores the user's own
    /// wallpaper instead of stranding them with a video frame.
    /// </summary>
    [JsonPropertyName("previousDesktopPictures")]
    public Dictionary<string, string> PreviousDesktopPictures { get; set; } = [];

    public WallpaperRecord? Wallpaper(Guid? id) =>
        id is null ? null : Library.FirstOrDefault(record => record.Id == id);

    public DisplayWallpaperConfiguration? Configuration(string displayId) =>
        DisplayConfigurations.FirstOrDefault(configuration => configuration.DisplayId == displayId);

    /// <summary>
    /// Inserts or replaces one display assignment, keeping the collection free
    /// of duplicates for the same display.
    /// </summary>
    public void SetConfiguration(DisplayWallpaperConfiguration configuration)
    {
        var index = DisplayConfigurations.FindIndex(existing => existing.DisplayId == configuration.DisplayId);
        if (index >= 0)
        {
            DisplayConfigurations[index] = configuration;
        }
        else
        {
            DisplayConfigurations.Add(configuration);
        }
    }

    /// <summary>
    /// Drops assignments that point at a wallpaper that is no longer in the
    /// library, so a deleted video cannot resurrect a broken session.
    /// </summary>
    public void RemoveAssignments(Guid wallpaperId)
    {
        foreach (var configuration in DisplayConfigurations.Where(item => item.WallpaperId == wallpaperId))
        {
            configuration.WallpaperId = null;
        }
    }

    /// <summary>Repairs anything a hand-edited or partially written file could carry.</summary>
    public void Normalize()
    {
        Version = CurrentVersion;
        Library ??= [];
        DisplayConfigurations ??= [];
        EnergyPreferences ??= new EnergyPreferences();
        PreviousDesktopPictures ??= [];

        Library.RemoveAll(record => string.IsNullOrEmpty(record.RelativePath));
        DisplayConfigurations.RemoveAll(configuration => string.IsNullOrEmpty(configuration.DisplayId));

        foreach (var configuration in DisplayConfigurations)
        {
            configuration.Normalize();
        }
    }
}
